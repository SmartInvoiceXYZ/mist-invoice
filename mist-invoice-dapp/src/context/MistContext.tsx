import React, { createContext, useContext, useEffect, useState } from "react";

import { StandardMerkleTree } from "@openzeppelin/merkle-tree";
import skiffCrypto from "@skiff-org/skiff-crypto";
import { BytesLike } from "ethers";
import randombytes from "randombytes";
import { useIndexedDB } from "react-indexed-db-hook";
import { logDebug, logError } from "../utils";
import { Web3Context } from "./Web3Context";

export type MistData = {
  merkleRoot?: BytesLike;
  clientRandom: BytesLike;
  providerRandom: BytesLike;
  clientKey: BytesLike;
  providerKey: BytesLike;
};

export type MistContextType = {
  data?: MistData;
  loading: boolean;
  getClientProof: () => BytesLike[] | undefined;
};

export const MistContext = createContext<MistContextType>({
  getClientProof: () => undefined,
  loading: true
});

export const MistContextProvider: React.FC<React.PropsWithChildren> = ({
  children
}) => {
  const { getByID, add } = useIndexedDB(process.env.MIST_DB_NAME || "");
  const { account } = useContext(Web3Context);
  const [data, setData] = useState<MistData>();
  const [tree, setTree] =
    useState<StandardMerkleTree<(BytesLike | undefined)[]>>();
  const [loading, setLoading] = useState(false);

  const providerAddress = process.env.MIST_PROVIDER_ADDRESS || "";

  useEffect(() => {
    if (!account || !providerAddress || !data) return;
    const values = [
      [account, data.clientRandom],
      [providerAddress, data.providerRandom]
    ];

    const _tree = StandardMerkleTree.of(values, ["address", "bytes32"]);
    setTree(_tree);
    setData({
      ...data,
      merkleRoot: _tree.root
    });
  }, [account, providerAddress, data]);

  const generate = () => {
    const clientRandom = randombytes(32);
    const providerRandom = randombytes(32);
    const clientKey = skiffCrypto.generateSymmetricKey();
    const providerKey = skiffCrypto.generateSymmetricKey();

    const generated = {
      clientRandom,
      providerRandom,
      clientKey,
      providerKey
    };

    setData(generated);

    return generated;
  };

  useEffect(() => {
    if (account) {
      setLoading(true);
      try {
        getByID<MistData>(account).then((res) => {
          if (res) {
            logDebug("Using existing data");
            setData(res);
          } else {
            logDebug("Generating new data");
            res = generate();
            add(res, account).then(() => {
              setData(res);
            });
          }
        });
      } catch (e) {
        logError(e);
      } finally {
        setLoading(false);
      }
    } else {
      setData(undefined);
    }
  }, [account]);

  const getClientProof = () => {
    if (account && tree && data) {
      const proof = tree.getProof([account, data.clientRandom]);
      return proof;
    }
    return undefined;
  };

  return (
    <MistContext.Provider value={{ data, loading, getClientProof }}>
      {children}
    </MistContext.Provider>
  );
};
