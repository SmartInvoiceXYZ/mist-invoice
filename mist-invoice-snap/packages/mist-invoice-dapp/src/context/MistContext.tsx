import React, { createContext, useContext, useEffect, useState } from 'react';

import { TransferType, UTXONote } from '../mist-sdk';
import {
  IncrementalMerkleTree,
  MerkleProof,
} from '@zk-kit/incremental-merkle-tree';
import { BytesLike, ethers, getBigInt, keccak256 } from 'ethers';
import { poseidon2 } from 'poseidon-lite';
import { randomBytes } from 'crypto';
import { useIndexedDB } from '../indexeddb';
import { logDebug, logError } from '../utils';
import { Web3Context } from './Web3Context';
import { connectSnap, getSnap, sendHello } from '@/snap/utils';
import {
  MetaMaskContext,
  MetamaskActions,
  MetamaskState,
} from './MetamaskContext';
import { DBConfig } from '@/dbconfig';

export type MistData = {
  merkleRoot?: BytesLike;
  clientRandom?: BytesLike;
  providerRandom?: BytesLike;
  clientKey?: BytesLike;
  providerKey?: BytesLike;
};

export type MistSecret = {
  merkleRoot?: bigint;
  providerHash: bigint;
  clientHash: bigint;
  encData: BytesLike[];
};

const utf8Encode = new TextEncoder();

export const SNARK_SCALAR_FIELD =
  21888242871839275222246405745257275088548364400416034343698204186575808495617n;
export const CLIENT_SIGNAL =
  getBigInt(keccak256(utf8Encode.encode('client'))) % SNARK_SCALAR_FIELD;
export const PROVIDER_SIGNAL =
  getBigInt(keccak256(utf8Encode.encode('provider'))) % SNARK_SCALAR_FIELD;
export const NULLIFYING_KEY =
  getBigInt(keccak256(utf8Encode.encode('nullifier'))) % SNARK_SCALAR_FIELD;

export type MistContextType = {
  data?: MistData;
  secret?: MistSecret;
  loading: boolean;
  amounts?: bigint[];
  receiver?: string;
  tokenAddress?: string;
  getClientProof: () => MerkleProof | undefined;
  handleSnapConnectClick: () => void;
  setAmounts: (amounts: bigint[]) => void;
  setReceiver: (receiver: string) => void;
  setTokenAddress: (tokenAddress: string) => void;
  state?: MetamaskState;
};

export const MistContext = createContext<MistContextType>({
  handleSnapConnectClick: () => undefined,
  getClientProof: () => undefined,
  loading: true,
  setAmounts: () => {},
  setReceiver: () => {},
  setTokenAddress: () => {},
});

export const MistContextProvider: React.FC<React.PropsWithChildren> = ({
  children,
}) => {
  const { getByID, add } = useIndexedDB(DBConfig.name, DBConfig);
  const { account } = useContext(Web3Context);
  const [data, setData] = useState<MistData>();
  const [secret, setSecret] = useState<MistSecret>();
  const [tree, setTree] = useState<IncrementalMerkleTree>();
  const [loading, setLoading] = useState(false);
  const [tokenAddress, setTokenAddress] = useState<string>('');
  const [receiver, setReceiver] = useState<string>('');
  const [amounts, setAmounts] = useState<bigint[]>([]);

  const providerAddress = process.env.MIST_PROVIDER_ADDRESS || '';

  // FROM MM SNAP TEMPLATE SITE
  const [state, dispatch] = useContext(MetaMaskContext);

  const handleSnapConnectClick = async () => {
    try {
      await connectSnap();
      const installedSnap = await getSnap();

      dispatch({
        type: MetamaskActions.SetInstalled,
        payload: installedSnap,
      });
    } catch (e) {
      console.error(e);
      dispatch({ type: MetamaskActions.SetError, payload: e });
    }
  };

  const handleSendHelloClick = async () => {
    try {
      await sendHello();
    } catch (e) {
      console.error(e);
      dispatch({ type: MetamaskActions.SetError, payload: e });
    }
  };
  //////////////////////////////////////////////////////////

  useEffect(() => {
    if (!account || !providerAddress || !data) return;
    const values = [
      [account, data.clientRandom],
      [providerAddress, data.providerRandom],
    ];

    const _tree = new IncrementalMerkleTree(poseidon2, 2, 0, 2, values);
    setTree(_tree);
    setData({
      ...data,
      merkleRoot: _tree.root,
    });
  }, [account, providerAddress, data]);

  useEffect(() => {
    const generate = async () => {
      const random = getBigInt(randomBytes(32).toString('hex'));
      // Contains bytes[] of encrypted note, clientKey and providerKey for each amount
      const encData = await Promise.all(
        amounts.map(async (amount, i) => {
          const note = new UTXONote({
            sender: account || '',
            receiver: receiver || '',
            amount,
            token: tokenAddress,
            identifier: BigInt(i),
            random,
            transferType: TransferType.Deposit,
            nullifyingKey: NULLIFYING_KEY,
          });
          const encryptedNote = await note.encrypt('goerli');
          return ethers.AbiCoder.defaultAbiCoder().encode(
            [
              'tuple(string encryptedData, string encryptedSenderKey, string encryptedReceiverKey)',
            ],
            [encryptedNote],
          );
        }),
      );
      const clientId = getBigInt(
        keccak256(
          ethers.AbiCoder.defaultAbiCoder().encode(['address'], [account]),
        ),
      );
      const providerId = getBigInt(
        keccak256(
          ethers.AbiCoder.defaultAbiCoder().encode(
            ['address'],
            [providerAddress],
          ),
        ),
      );
      const clientHash = poseidon2([
        poseidon2([clientId, NULLIFYING_KEY]),
        random,
      ]);
      const providerHash = poseidon2([
        poseidon2([providerId, NULLIFYING_KEY]),
        random,
      ]);

      const newSecret = {
        providerHash,
        clientHash,
        encData,
      };

      setSecret(newSecret);
      return secret;
    };

    async function genAddRes() {
      let res = await generate();
      add(res, account).then(() => {
        setSecret(res);
      });
    }

    if (account) {
      setLoading(true);
      try {
        getByID(account).then((res) => {
          if (res) {
            logDebug('Using existing data');
            setData(res);
          } else {
            logDebug('Generating new data');
            genAddRes();
          }
        });
      } catch (e) {
        logError(e);
      } finally {
        setLoading(false);
      }
    } else {
      setData(undefined);
    }
  }, [account, receiver, tokenAddress, amounts]);

  const getClientProof = () => {
    if (account && tree && data) {
      const proof = tree.createProof(
        tree.indexOf(poseidon2([getBigInt(account), CLIENT_SIGNAL])),
      );
      return proof;
    }
    return undefined;
  };

  return (
    <MistContext.Provider
      value={{
        data,
        secret,
        amounts,
        receiver,
        tokenAddress,
        loading,
        getClientProof,
        handleSnapConnectClick,
        setAmounts,
        setReceiver,
        setTokenAddress,
        state,
      }}
    >
      {children}
    </MistContext.Provider>
  );
};
