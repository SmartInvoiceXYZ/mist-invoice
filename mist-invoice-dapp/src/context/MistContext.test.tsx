import { render } from "@testing-library/react";
import { useContext } from "react";

import React from "react";
import { MistContext, MistContextProvider } from "./MistContext";

jest.mock("@gnosis.pm/safe-apps-web3modal");
jest.mock("@walletconnect/web3-provider");
jest.mock("web3");

jest.mock("react-indexed-db-hook", () => ({
  useIndexedDB: jest.fn().mockImplementation(() => ({
    getByID: jest.fn(),
    add: jest.fn()
  }))
}));

const TestingComponent = () => {
  const { data, loading } = useContext(MistContext);

  return (
    <div>
      {loading ? (
        <div data-testid="div:loader">Loading...</div>
      ) : data ? (
        <div data-testid="div:data">
          <p data-testid="p:merkle-root">Merkle Root: {data.merkleRoot}</p>
          <p data-testid="p:client-random">
            Client Random: {data.clientRandom}
          </p>
          <p data-testid="p:client-key">Client Key: {data.clientKey}</p>
          <p data-testid="p:provider-random">
            Provider Random: {data.providerRandom}
          </p>
          <p data-testid="p:provider-key">
            Provider Random: {data.providerKey}
          </p>
        </div>
      ) : (
        <div data-testid="div:data">No data</div>
      )}
    </div>
  );
};

describe("MistContext", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it("should render", () => {
    const view = render(
      <MistContextProvider>
        <TestingComponent />
      </MistContextProvider>
    );
    expect(view.asFragment()).toMatchSnapshot();
  });
});
