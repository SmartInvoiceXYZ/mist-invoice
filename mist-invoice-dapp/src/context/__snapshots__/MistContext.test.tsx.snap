// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MistContext should render 1`] = `
<DocumentFragment>
  <div>
    <div
      data-testid="div:data"
    >
      No data
    </div>
  </div>
</DocumentFragment>
`;
